src/redisgears_python.c

```diff
@@ -3317,27 +3333,108 @@ static PyObject* scriptRunnerRun(PyObject *cls, PyObject *args){
         PyErr_SetString(GearsError, "Given argument is not of type PyTorchScriptRunner");
         return NULL;
     }
+	if(pys->s == NULL){
+		PyErr_SetString(GearsError, "PyTorchScriptRunner is invalid");
+		return NULL;
+	}
     RAI_Error* err;
     RedisAI_InitError(&err);
+	RAI_ScriptRunCtx *sctx = pys->s;
+	pys->s = NULL;
     PyThreadState* _save = PyEval_SaveThread();
-    RedisAI_ScriptRun(pys->s, err);
+    RedisAI_ScriptRun(sctx, err);
     PyEval_RestoreThread(_save);
     if (RedisAI_GetErrorCode(err) != RedisAI_ErrorCode_OK) {
         PyErr_SetString(GearsError, RedisAI_GetError(err));
         RedisAI_FreeError(err);
+        RedisAI_ScriptRunCtxFree(sctx);
         return NULL;
     }
     RedisAI_FreeError(err);
     PyObject* tensorList = PyList_New(0);
-    for(size_t i = 0 ; i < RedisAI_ScriptRunCtxNumOutputs(pys->s) ; ++i){
+    for(size_t i = 0 ; i < RedisAI_ScriptRunCtxNumOutputs(sctx) ; ++i){
         PyTensor* pyt = PyObject_New(PyTensor, &PyTensorType);
-        pyt->t = RedisAI_TensorGetShallowCopy(RedisAI_ScriptRunCtxOutputTensor(pys->s, i));
+        pyt->t = RedisAI_TensorGetShallowCopy(RedisAI_ScriptRunCtxOutputTensor(sctx, i));
         PyList_Append(tensorList, (PyObject*)pyt);
         Py_DECREF(pyt);
     }
+	RedisAI_ScriptRunCtxFree(sctx);
     return tensorList;
 }
 
+static void FinishAsyncScriptRun(RAI_OnFinishCtx *onFinishCtx, void *private_data) {
+
+	RAI_Error *error;
+	RedisAI_InitError(&error);
+	RAI_ScriptRunCtx* sctx = RedisAI_GetAsScriptRunCtx(onFinishCtx, error);
+
+	RedisGearsPy_LOCK
+	PyObject* future = private_data;
+	PyObject* tensorList = PyList_New(0);
+	PyObject* pArgs = PyTuple_New(2);
+	PyTuple_SetItem(pArgs, 0, future);
+
+	if (RedisAI_GetErrorCode(error) != RedisAI_ErrorCode_OK) {
+		const char *errStr = RedisAI_GetError(error);
+		PyObject* pyErr = PyUnicode_FromStringAndSize(errStr, strlen(errStr));
+		PyTuple_SetItem(pArgs, 1, pyErr);
+		PyObject_CallObject(setFutureExceptionFunction, pArgs);
+		goto finish;
+	}
+
+	for(size_t i = 0 ; i < RedisAI_ScriptRunCtxNumOutputs(sctx) ; ++i){
+		PyTensor* pyt = PyObject_New(PyTensor, &PyTensorType);
+		pyt->t = RedisAI_TensorGetShallowCopy(RedisAI_ScriptRunCtxOutputTensor(sctx, i));
+		PyList_Append(tensorList, (PyObject*)pyt);
+		Py_DECREF(pyt);
+	}
+
+	PyTuple_SetItem(pArgs, 1, tensorList);
```

Comment:
Again I am not sure why setting the list inside a tuple is needed instead if just passing the list (again also apply to ModerRunAsync)